common/src/Model/Brush.cpp

```diff
@@ -969,20 +1007,26 @@ namespace TrenchBroom {
                 swap(remaining, moving);
                 delta = -delta;
             }
-            
-            // Now check if any of the moving vertices would travel into or through the remaining fragment.
+
+            // Now check if any of the moving vertices would travel through the remaining fragment.
             for (const BrushVertex* vertex : moving.vertices()) {
                 const Vec3& oldPos = vertex->position();
                 const Vec3 newPos = oldPos + delta;
                 
+                // Skip moving vertices that end up inside the remaining fragment.
+                // These are handled in the allowVertexRemoval section above.
+                if (remaining.contains(newPos))
+                    continue;
+                
                 for (const BrushFaceGeometry* face : remaining.faces()) {
                     if (face->pointStatus(oldPos) == Math::PointStatus::PSAbove &&
```

Comment:
I think you're testing for the wrong faces now. If you modify these lines to

```
if (face->pointStatus(oldPos) == Math::PointStatus::PSBelow &&
    face->pointStatus(newPos) == Math::PointStatus::PSAbove) {
```

then you'll only consider faces such that the vertex travelled from behind to in front of it. I think that it should be enough to check whether the ray on which the vertex travels actually intersected with the face's polygon.